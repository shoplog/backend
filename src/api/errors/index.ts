export * from './http.error';
export * from './unauthorized.error';
