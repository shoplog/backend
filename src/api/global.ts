import { VPICController } from 'src/api/controllers/vpic.controller';
import { VPICDatabase } from 'src/common/initializers/database';
import {
	LookupRepository,
	MakeRepository,
	ModelRepository,
	VinRepository,
	YearRepository,
} from 'src/data/vpic/repositories';
import { VPICService } from 'src/domains/vpic/services/vpic.service';

const vinRepository = new VinRepository(VPICDatabase);
const yearRepository = new YearRepository(VPICDatabase);
const makeRepository = new MakeRepository(VPICDatabase);
const modelRepository = new ModelRepository(VPICDatabase);
const lookupRepository = new LookupRepository(VPICDatabase);
const vpicService = new VPICService(vinRepository, yearRepository, makeRepository, modelRepository, lookupRepository);
const vpicController = new VPICController(vpicService);

export { vpicController };
