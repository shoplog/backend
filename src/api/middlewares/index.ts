export * from './error-handler.middleware';
export * from './contex-wrap.middleware';
export * from './check-jwt.middleware';
