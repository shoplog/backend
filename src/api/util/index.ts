export * from 'src/api/util/helpers';
