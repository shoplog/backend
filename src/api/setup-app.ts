import { apiReference } from '@scalar/express-api-reference';
import compression from 'compression';
import express, { Request, json } from 'express';
import { middleware as OpenApiValidatorMiddleware } from 'express-openapi-validator';
import { NotFound } from 'express-openapi-validator/dist/openapi.validator';
import helmet from 'helmet';
import pinoHttp from 'pino-http';
import { CUSTOM_HEADERS } from 'src/api/constants/headers';
import { errorHandlerMiddleware } from 'src/api/middlewares';
import { createVehiclesRoutes } from 'src/api/routes/vehicles.route';
import { logger } from 'src/common/initializers/logger';

const apiSpec = 'data/openapi/v1.yml';

export const setupApp = async () => {
	const app = express();

	app.use(
		helmet({
			// Disable CSP because it's not needed for this API only project.
			contentSecurityPolicy: false,
		})
	);

	// For when the app is behind a proxy, like in a docker container or in a kubernetes cluster.
	app.set('trust proxy', 1);

	app.use(compression());

	app.use(
		'/api-docs',
		apiReference({
			spec: {
				content: apiSpec,
			},
		})
	);

	// Generate request id that will correlate all logs for a single request.
	app.use((req, res, next) => {
		req.id = crypto.randomUUID();
		res.set(CUSTOM_HEADERS.RequestId, req.id);
		next();
	});

	app.use(
		pinoHttp({
			logger,
			quietReqLogger: true,
			genReqId: (req) => req.id,
		})
	);

	app.use(json());

	app['get']('/', (req, res) => {
		res.status(200).end();
	});

	app.use('/api/v1', await createVehiclesRoutes());

	app.use(
		OpenApiValidatorMiddleware({
			apiSpec,
			validateApiSpec: false,
			validateRequests: true,
			validateResponses: true,
		})
	);

	app.use((req: Request) => {
		throw new NotFound({ path: req.url, message: 'Not Found' });
	});

	app.use(errorHandlerMiddleware());

	return app;
};
