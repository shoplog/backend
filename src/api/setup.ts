import { apiReference } from '@scalar/express-api-reference';
import compression from 'compression';
import cors from 'cors';
import express, { json } from 'express';
import { middleware as OpenApiValidatorMiddleware } from 'express-openapi-validator';
import { OpenApiSpecLoader } from 'express-openapi-validator/dist/framework/openapi.spec.loader';
import { NotFound, OpenAPIV3 } from 'express-openapi-validator/dist/framework/types';
import 'express-openapi-validator/dist/middlewares/parsers/schema.parse';
import helmet from 'helmet';
import pinoHttp from 'pino-http';
import { OPEN_API_SPEC_FILE_PATH } from 'src/api/constants/files';
import { CUSTOM_HEADERS } from 'src/api/constants/headers';
import { Dependencies } from 'src/api/global';
import { errorHandlerMiddleware } from 'src/api/middlewares';
import { createVPICRoutes } from 'src/api/routes/vpic.routes';
import { logger } from 'src/common/initializers/logger';

export const loadOpenApiSpec = async (): Promise<OpenAPIV3.Document> => {
	const loader = new OpenApiSpecLoader({
		apiDoc: OPEN_API_SPEC_FILE_PATH,
		validateApiSpec: true,
	});

	const { apiDoc: apiSpec } = await loader.load();

	return apiSpec;
};

export const createApp = async (dependencies: Dependencies) => {
	const app = express();
	const apiSpec = await loadOpenApiSpec();

	app.use(
		helmet({
			// Disable CSP because it's not needed for this API only project.
			contentSecurityPolicy: false,
		})
	);

	// For when the app is behind a proxy, like in a docker container or in a kubernetes cluster.
	app.set('trust proxy', 1);

	app.use(compression());

	app.use(cors());

	app.get('/docs/openapi.json', (_req, res) => {
		res.json(apiSpec);
	});

	app.use(
		'/docs',
		apiReference({
			spec: {
				content: apiSpec,
			},
		})
	);

	// Generate request id that will correlate all logs for a single request.
	app.use((req, res, next) => {
		req.id = crypto.randomUUID();
		res.set(CUSTOM_HEADERS.RequestId, req.id);
		next();
	});

	app.use(
		pinoHttp({
			logger,
			quietReqLogger: true,
			genReqId: (req) => req.id,
		})
	);

	app.use(json());

	app.use(
		OpenApiValidatorMiddleware({
			apiSpec,
			validateApiSpec: true,
			validateRequests: true,
			validateResponses: true,
		})
	);

	app['get']('/', (req, res) => {
		res.status(200).end();
	});

	app.use('/api/v1', await createVPICRoutes(dependencies.controllers.vpicController));

	app.use((req, _res, _next) => {
		throw new NotFound({ path: req.originalUrl });
	});

	app.use(errorHandlerMiddleware());

	return app;
};
