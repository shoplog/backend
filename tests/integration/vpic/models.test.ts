import { Express } from 'express';
import { createApp } from 'src/api/setup';
import supertest from 'supertest';

const baseUrl = '/api/v1/vpic';

describe('/vpic/models', () => {
	let app: Express;
	const resourceUrl = '/api/v1/vpic/models';

	beforeAll(async () => {
		app = await createApp();
	});

	describe('GET /vpic/models/:modelId/year/:year/attributes', () => {
		it('should respond with 200 Ok - attributes for model and year', async () => {
			// Arrange
			const modelId = 2223; // Tacoma
			const year = 2002;

			// Act
			const body = await supertest(app)
				.get(`${resourceUrl}/${modelId}/year/${year}/attributes`)
				.expect(200)
				.then((res) => res.body);

			// Assert
			expect(body).toMatchSnapshot();
		});
	});
});
