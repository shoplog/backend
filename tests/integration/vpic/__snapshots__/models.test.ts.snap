// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/vpic/models GET /vpic/models/:modelId/year/:year/attributes should respond with 200 Ok - attributes for model and year 1`] = `
[
  {
    "code": "bodyCabType",
    "description": "Cab type applies to both pickup truck and other medium- and heavy-duty trucks. The cab or cabin of a truck is the inside space in a truck where the driver is seated. For pickup trucks, the cab type is categorized by the combination number of doors and number of rows for seating. For medium- and heavy-duty trucks (MDHD), the cab type is categorized by the relative location of engine and cab. For pickup trucks, there are four cab types. Regular: 2 door, 1 row of seats Extra/Super/Quad/Double/King/Extended: 2 doors, 2 rows of seats Crew/Super Crew/Crew Max: 4 doors, 2 rows of seats Mega: 4 doors, 2 rows of seats (with a bigger cabin than crew cab type) For medium- and heavy-duty (MDHD) trucks, there are several categories as listed below. Cab Beside Engine CAE: Cab Above Engine CBE: Cab Behind Engine COE: Cab Over Engine or Flat Nose: Driver sits on top of the front axle and engine LCF: Low Cab Forward Conventional: Driver sits behind the engine Non-Tilt Tilt",
    "name": "Cab Type",
    "values": [
      {
        "id": 3,
        "value": "Regular",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": 2,
        "value": "Extra/Super/Quad/Double/King/Extended",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "bodyClass",
    "description": "Body Class presents the body type based on 49 CFR 565.12(b): "Body type means the general configuration or shape of a vehicle distinguished by such characteristics as the number of doors or windows, cargo-carrying features and the roofline (e.g., sedan, fastback, hatchback)." Definitions are not provided for individual body types in the regulation.",
    "name": "Body Class",
    "values": [
      {
        "id": 60,
        "value": "Pickup",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "brakeSystemType",
    "description": "Brake system type is the type of brake systems used to stop and hold the vehicle or combination of motor vehicles.",
    "name": "Brake System Type",
    "values": [
      {
        "id": 2,
        "value": "Hydraulic",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "displacementL",
    "description": "Engine displacement (in liters) is the volume swept by all the pistons inside the cylinders of a reciprocating engine in a single movement from top dead center to bottom dead center.",
    "name": "Displacement (L)",
    "values": [
      {
        "id": null,
        "value": 2.4,
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": 2.7,
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": 3.4,
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "driveType",
    "description": "Drive type stores information about vehicle drivetrain configuration. The most common drive types for passenger cars, crossover vehicles, and pickup trucks are front-wheel drive (FWD), rear-wheel drive (RWD), all-wheel drive (AWD), and 4-wheel drive (4WD).",
    "name": "Drive Type",
    "values": [
      {
        "id": 7,
        "value": "4x2",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": 2,
        "value": "4WD/4-Wheel Drive/4x4",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "engineConfiguration",
    "description": "Engine configuration defines how engine cylinders are arranged. Common values are V6 for V-shaped arrangement, I4 or L4 for in-line arrangement.",
    "name": "Engine Configuration",
    "values": [
      {
        "id": 1,
        "value": "In-Line",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": 2,
        "value": "V-Shaped",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "engineCylinders",
    "description": "This is a numerical field to store the number of cylinders in an engine. Common values for passenger cars are 4 or 6.",
    "name": "Engine Number of Cylinders",
    "values": [
      {
        "id": null,
        "value": 6,
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": 4,
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "engineHp",
    "description": "Engine brake is the horsepower (hp) at the engine output shaft. Engine Brake (hp) From is the lower value of the range.",
    "name": "Engine Brake (hp) From",
    "values": [
      {
        "id": null,
        "value": 190,
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": 150,
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": 142,
        "vinSchemaIds": [
          6869,
        ],
      },
    ],
  },
  {
    "code": "engineModel",
    "description": "Engine model is a name that a manufacturer applies to a family of engine.",
    "name": "Engine Model",
    "values": [
      {
        "id": null,
        "value": "5VZ-FE",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": "3RZ-FE",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": "2RZ-FE",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "fuelTypePrimary",
    "description": "Fuel type defines the fuel used to power the vehicle. For vehicles that have two power sources, such as plug-in hybrid vehicle, both primary fuel type and secondary fuel type will be provided.",
    "name": "Fuel Type - Primary",
    "values": [
      {
        "id": 4,
        "value": "Gasoline",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "gvwr",
    "description": "Gross vehicle weight rating (GVWR) is the maximum operating weight of a vehicle including the vehicle's chassis, body, engine, engine fluids, fuel, accessories, driver, passengers and cargo, but excluding that of the trailers. Per 49 CFR 565.15, Class 1 is further broken down to Class A-D; Class 2 is further broken down to Class E-H. This field captures the lower bound of GVWR range for the vehicle.",
    "name": "Gross Vehicle Weight Rating From",
    "values": [
      {
        "id": 13,
        "value": "Class 1D: 5,001 - 6,000 lb (2,268 - 2,722 kg)",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": 12,
        "value": "Class 1C: 4,001 - 5,000 lb (1,814 - 2,268 kg)",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "model",
    "description": "Per 49 CFR 565, Model means a name that a manufacturer applies to a family of vehicles of the same type, make, line, series and body type.",
    "name": "Model",
    "values": [
      {
        "id": 2223,
        "value": "Tacoma",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "ncsaBodyType",
    "description": "An internal NHTSA field to capture the body type of the vehicle.",
    "name": "NCSA Body Type",
    "values": [
      {
        "id": 34,
        "value": "Light Pickup",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "ncsaMake",
    "description": "An internal NHTSA field to capture the Make of the vehicle.",
    "name": "NCSA Make",
    "values": [
      {
        "id": 49,
        "value": "Toyota",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "ncsaModel",
    "description": "An internal NHTSA field to capture the Model of the vehicle.",
    "name": "NCSA Model",
    "values": [
      {
        "id": 49000472,
        "value": "Tacoma",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "note",
    "description": "Note is used to store any additional information that does not correspond to any of the specified fields on the interface. This is a catch-all element for systems other than for engine, restraint system, brake and battery. Engine, restraint system, brake, and battery have their own note elements.",
    "name": "Note",
    "values": [
      {
        "id": null,
        "value": "Chassis: With 1/2 ton Independent Front Suspension",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": "Extra Cab",
        "vinSchemaIds": [
          6869,
        ],
      },
      {
        "id": null,
        "value": "Double Cab",
        "vinSchemaIds": [
          6869,
        ],
      },
      {
        "id": null,
        "value": "Standard Cab",
        "vinSchemaIds": [
          6869,
        ],
      },
    ],
  },
  {
    "code": "plantCity",
    "description": "This data element captures the city of the manufacturing plant where the manufacturer affixes the VIN.",
    "name": "Plant City",
    "values": [
      {
        "id": null,
        "value": "FREMONT",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "plantCompanyName",
    "description": "This data element captures the name of the company that owns the manufacturing plant  where the manufacturer affixes the VIN.",
    "name": "Plant Company Name",
    "values": [
      {
        "id": null,
        "value": "New United Motor Manufacturing (NUMMI)",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "plantCountry",
    "description": "This data element captures the country of the manufacturing plant where the manufacturer affixes the VIN.",
    "name": "Plant Country",
    "values": [
      {
        "id": 6,
        "value": "UNITED STATES (USA)",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "plantState",
    "description": "This data element captures the State or Province name within the Plant Country of the manufacturing plant where the manufacturer affixes the VIN.",
    "name": "Plant State",
    "values": [
      {
        "id": null,
        "value": "CALIFORNIA",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "series",
    "description": "Per 49 CFR 565, Series means a name that a manufacturer applies to a subdivision of a "line" denoting price, size or weight identification and that is used by the manufacturer for marketing purposes.",
    "name": "Series",
    "values": [
      {
        "id": null,
        "value": "RZN191L/RZN196L/VZN195L",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": "RZN171L/VZN170L",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": "RZN140L",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": "RZN150L/VZN150L",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": null,
        "value": "RZN161L/VZN160L",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "trim",
    "description": "Trim levels further identify a vehicle by a particular set of special features. Higher trim levels either will add to the features of the base (entry-level model), or replace them with something else.",
    "name": "Trim",
    "values": [
      {
        "id": null,
        "value": "SR5",
        "vinSchemaIds": [
          5691,
        ],
      },
      {
        "id": null,
        "value": "DELUXE",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
  {
    "code": "wheelBaseType",
    "description": "This field describes the wheel base variations for some trucks and passenger cars, relative to other variants of the vehicle model. This field may have values such as short, standard, long, extra long, or super long.",
    "name": "Wheel Base Type",
    "values": [
      {
        "id": 4,
        "value": "Extra Long",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
      {
        "id": 2,
        "value": "Short",
        "vinSchemaIds": [
          5691,
          6869,
        ],
      },
    ],
  },
]
`;
